In [1]:
from geopy.geocoders import Nominatim

In [2]:
geolocator = Nominatim()

In [3]:
location = geolocator.geocode("175 5th Avenue NYC")
print(location.address)

Flatiron Building, 175, 5th Avenue, Flatiron, New York, NYC, New York, 10010, United States of America


In [4]:
print location.address

Flatiron Building, 175, 5th Avenue, Flatiron, New York, NYC, New York, 10010, United States of America


In [5]:
print((location.latitude, location.longitude))

(40.7410861, -73.9896297241625)


In [7]:
print location.latitude, location.longitude

40.7410861 -73.9896297242


In [8]:
import pandas as pd
import numpy as np

In [9]:
pwd

u'C:\\Users\\andersrmr\\Documents\\Python Scripts\\galvanize_prep\\pothole_repair\\sandbox'

In [10]:
df = pd.read_csv('data/Pothole_Repairs_Seattle.csv')

In [11]:
df.ix[0,'ADDRDESC'], df.ix[0,'LOCATION']

('HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYON ST                                                                                                                                                                                                             ',
 'Holden pl SW bt  SW Kenyon Holden                                                                                                                                                                                                                              ')

In [12]:
df.ix[0, 'ADDRDESC']

'HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYON ST                                                                                                                                                                                                             '

In [14]:
geolocator.geocode(df.ix[0, 'ADDRDESC']) #gives nothing at all

In [15]:
https://maps.google.com/maps?q=Commonwealth+Avenue+at+Washington+street,+Boston,+MA&hl=en&sll=42.351351,-71.118573&sspn=0.054868,0.132093&t=h&hnear=Commonwealth+Ave+%26+Washington+St,+Boston,+Suffolk,+Massachusetts+02135&z=17

SyntaxError: invalid syntax (<ipython-input-15-0482f57ad101>, line 1)

In [21]:
%wget http://maps.googleapis.com/maps/api/geocode/json?address=market+and+4th,+san+francisco&sensor=false

ERROR: Line magic function `%wget` not found.


In [22]:
from geopy.geocoders import GoogleV3
geolocator = GoogleV3()

In [23]:
address, (latitude, longitude) = geolocator.geocode("market and 4th san francisco")

In [24]:
print address

Market St & 4th St, San Francisco, CA 94102, USA


In [25]:
print (latitude, longitude)

(37.7857158, -122.4059115)


In [26]:
address, (latitude, longitude) = geolocator.geocode(df.ix[0, 'ADDRDESC'])

TypeError: 'NoneType' object is not iterable

In [27]:
address, (latitude, longitude) = geolocator.geocode('SW HOLDEN ST AND SW KENYON ST seattle')

In [28]:
print address, (latitude, longitude)

SW Kenyon St & Holden Pl SW, Seattle, WA 98126, USA (47.5318997, -122.3717987)


In [29]:
df.ix[0,'ADDRDESC'], df.ix[0,'LOCATION']

('HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYON ST                                                                                                                                                                                                             ',
 'Holden pl SW bt  SW Kenyon Holden                                                                                                                                                                                                                              ')

In [30]:
df.ix[1,'ADDRDESC'], df.ix[1,'LOCATION']

('MONTLAKE BLVD NE AND NE PACIFIC ST                                                                                                                                                                                                                             ',
 'NE Pacific Street, one block west of Montlake Boulevard N                                                                                                                                                                                                      ')

In [31]:
df.columns

Index([u'OBJECTID', u'WOKEY', u'WO_STATUS', u'LAST90DAYRPRD', u'CURRENT_QTR',
       u'PREVIOUS_1QTR', u'PREVIOUS_2QTR', u'PREVIOUS_3QTR', u'PREVIOUS_4QTR',
       u'INITDT', u'FLDSTARTDT', u'FLDENDDT', u'LOCATION', u'ADDRDESC',
       u'SHAPE'],
      dtype='object')

In [34]:
latitude, longitude = geolocator.geocode('montlake blvd NE and NE pacific st seattle')
print latitude, longitude

Montlake Blvd NE & NE Pacific St, Seattle, WA 98105, USA (47.6490137, -122.3046552)


In [35]:
geolocator.geocode('montlake blvd NE and NE pacific st seattle')

Location((47.6490137, -122.3046552, 0.0))

In [36]:
print latitude

Montlake Blvd NE & NE Pacific St, Seattle, WA 98105, USA


In [37]:
print longitude

(47.6490137, -122.3046552)


In [39]:
address, geocoords = geolocator.geocode('montlake blvd NE and NE pacific st seattle')
print address, geocoords

Montlake Blvd NE & NE Pacific St, Seattle, WA 98105, USA (47.6490137, -122.3046552)


In [41]:
print longitude[0]

47.6490137


In [42]:
df.ix[0:5,'ADDRDESC']

0    HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYO...
1    MONTLAKE BLVD NE AND NE PACIFIC ST            ...
2    NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY ...
3    S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE...
4    N 80TH ST BETWEEN CORLISS (W LEG) AVE N AND CO...
5    18TH AVE NE BETWEEN NE 52ND ST AND NE 55TH ST ...
Name: ADDRDESC, dtype: object

In [43]:
for idx in xrange(5):
    print df.ix[idx,'ADDRDESC']

HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYON ST                                                                                                                                                                                                             
MONTLAKE BLVD NE AND NE PACIFIC ST                                                                                                                                                                                                                             
NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY WAY NE                                                                                                                                                                                                           
S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE S                                                                                                                                                                                        

In [44]:
for idx in xrange(15):
    print df.ix[idx,'ADDRDESC']

HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYON ST                                                                                                                                                                                                             
MONTLAKE BLVD NE AND NE PACIFIC ST                                                                                                                                                                                                                             
NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY WAY NE                                                                                                                                                                                                           
S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE S                                                                                                                                                                                        

In [45]:
geolocator.geocode('HILLSIDE DR E AND LAKE WASHINGTON BLVD E')

Location((47.6224796, -122.2846636, 0.0))

In [46]:
geolocator.geocode('NE 45TH ST')

Location((47.661257, -122.2986688, 0.0))

In [47]:
geolocator.geocode ('NE 45TH ST Seattle')

Location((47.661257, -122.2986688, 0.0))

In [48]:
geolocator.geocode('HILLSIDE DR E AND LAKE WASHINGTON BLVD E seattle')

Location((47.6224796, -122.2846636, 0.0))

In [52]:
geolocator.geocode('HILLSIDE DR E AND LAKE WASHINGTON BLVD E seattle')[1]

(47.6224796, -122.2846636)

In [54]:
geolocator.geocode('NE 45TH ST BETWEEN 15TH AVE NE AND 16TH AVE NE seattle')

Location((47.6757999, -122.3108485, 0.0))

In [56]:
df.shape[0]

26738

In [63]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(df.shape[0]):
        loc = df.ix[row,'ADDRDESC']
        new_locs.append(geolocator.geocode(loc))
        return new_locs

In [68]:
get_geocodes(df)

[None]

In [70]:
type(df.ix[0,'ADDRDESC'])

str

In [78]:
geolocator.geocode(df.ix[1,'ADDRDESC'])

Location((47.6490137, -122.3046552, 0.0))

In [79]:

df.ix[0:5,'ADDRDESC']

0    HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYO...
1    MONTLAKE BLVD NE AND NE PACIFIC ST            ...
2    NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY ...
3    S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE...
4    N 80TH ST BETWEEN CORLISS (W LEG) AVE N AND CO...
5    18TH AVE NE BETWEEN NE 52ND ST AND NE 55TH ST ...
Name: ADDRDESC, dtype: object

In [ ]:
AIzaSyBpxr2Rx1WbYqrxpuUg_I819YuDw0Pvm_0                                                                                                                                                                                                                                                  

In [80]:
for idx in xrange(15):
    print df.ix[idx,'ADDRDESC']

HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYON ST                                                                                                                                                                                                             
MONTLAKE BLVD NE AND NE PACIFIC ST                                                                                                                                                                                                                             
NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY WAY NE                                                                                                                                                                                                           
S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE S                                                                                                                                                                                        

In [83]:
df.ix[0,'ADDRDESC'] + ' Seattle'

'HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYON ST                                                                                                                                                                                                              Seattle'

In [91]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(df.shape[0]):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc)))
    return new_locs
geo_codes = get_geocodes(df)

GeocoderQuotaExceeded: The given key has gone over the requests limit in the 24 hour period or has submitted too many requests in too short a period of time.

In [ ]:
print geo_codes[0:5]

In [92]:
geolocator = GoogleV3('AIzaSyBpxr2Rx1WbYqrxpuUg_I819YuDw0Pvm_0')

In [93]:
geolocator.geocode(df.ix[0,'ADDRDESC'] + ' Seattle')

Location((47.5318578, -122.3631857, 0.0))

In [98]:
def get_geocodes(df):
    new_locs = []
#     for row in xrange(df.shape[0]):
    for row in xrange(5):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc)))
#         new_locs.append(row)
#         new_locs.append(loc)
    return new_locs

In [99]:
get_geocodes(df)

[(0, Location((47.5318578, -122.3631857, 0.0))),
 (1, Location((47.6490137, -122.3046552, 0.0))),
 (2, Location((47.7074461, -122.3015663, 0.0))),
 (3, Location((47.5970845, -122.3250649, 0.0))),
 (4, Location((47.6062095, -122.3320708, 0.0)))]

In [102]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(50):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc)))
    return new_locs

In [103]:
geocodes_1to50 = get_geocodes(df)

In [104]:
print geocodes_1to50

[(0, Location((47.5318578, -122.3631857, 0.0))), (1, Location((47.6490137, -122.3046552, 0.0))), (2, Location((47.7074461, -122.3015663, 0.0))), (3, Location((47.5970845, -122.3250649, 0.0))), (4, Location((47.6062095, -122.3320708, 0.0))), (5, Location((47.6685216, -122.2956295, 0.0))), (6, Location((47.7214758, -122.3649839, 0.0))), (7, Location((47.5158683, -122.3551443, 0.0))), (8, Location((47.55013, -122.3763528, 0.0))), (9, Location((47.6231312, -122.3330033, 0.0))), (10, Location((47.6503107, -122.3886039, 0.0))), (11, Location((47.7043612, -122.3527632, 0.0))), (12, Location((47.610741, -122.338703, 0.0))), (13, Location((47.6757999, -122.3108485, 0.0))), (14, Location((47.6224796, -122.2846636, 0.0))), (15, Location((47.5682933, -122.3018407, 0.0))), (16, Location((47.7086636, -122.3273083, 0.0))), (17, Location((47.6062095, -122.3320708, 0.0))), (18, Location((47.7290258, -122.343668, 0.0))), (19, Location((47.7305514, -122.3381845, 0.0))), (20, Location((47.693551, -122.385

In [107]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(50,100):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc)))
    return new_locs

In [108]:
len(range(50,100))

50

In [109]:
geocodes_51to100 = get_geocodes(df)

In [115]:
geocodes_51to100[:10]

[(50, Location((47.6062095, -122.3320708, 0.0))),
 (51, Location((47.5509151, -122.2976812, 0.0))),
 (52, Location((47.7104453, -122.3313537, 0.0))),
 (53, Location((47.659178, -122.2900991, 0.0))),
 (54, Location((47.6340238, -122.3544058, 0.0))),
 (55, Location((47.6685216, -122.2956295, 0.0))),
 (56, Location((47.6231334, -122.326594, 0.0))),
 (57, Location((47.7212347, -122.3140735, 0.0))),
 (58, Location((47.666443, -122.3034441, 0.0))),
 (59, Location((47.6996341, -122.3473049, 0.0)))]

In [111]:
geocodes_51to100[-10:]

[(90, Location((47.6324363, -122.3556352, 0.0))),
 (91, Location((47.6297515, -122.3488124, 0.0))),
 (92, Location((47.6152216, -122.3261797, 0.0))),
 (93, Location((47.7160266, -122.3662914, 0.0))),
 (94, Location((47.5442208, -122.2797783, 0.0))),
 (95, Location((47.4966269, -122.276053, 0.0))),
 (96, Location((47.6062095, -122.3320708, 0.0))),
 (97, Location((47.669457, -122.2848171, 0.0))),
 (98, Location((47.7104453, -122.3313537, 0.0))),
 (99, Location((47.6208963, -122.3430313, 0.0)))]

In [112]:
df.shape[0]

26738

In [125]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(100,500):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc, timeout=10)))
    return new_locs

In [126]:
geocodes_101to500 = get_geocodes(df)

In [187]:
lats, longs = [], []
for ind in xrange(len(geocodes_101to500)):
    lats.append(geocodes_101to500[ind][1].latitude)
    longs.append(geocodes_101to500[ind][1].longitude)

In [188]:
len(lats), len(set(lats)), len(set(longs)) #why am I not getting 400 unique latitudes?

(400, 275, 275)

In [158]:
geocodes_101to500[:100][0][1]

Location((47.6981806, -122.3446, 0.0))

In [193]:
geocodes_101to500[0][1].latitude, geocodes_101to500[0][1].longitude

(47.6981806, -122.3446)

In [195]:
geocodes_1to500_lat = geocodes_1to50[0][1].latitude + geocodes_51to100[0][1].latitude + geocodes_101to500[0][1].latitude

In [196]:
geocodes_1to500_long = geocodes_1to50[0][1].longitude + geocodes_51to100[0][1].longitude + geocodes_101to500[0][1].longitude

In [206]:
geocodes_1to500 = geocodes_1to50 + geocodes_51to100 + geocodes_101to500

In [209]:
type(geocodes_1to500)

list

In [220]:
lats, longs = [], []
for elem in geocodes_1to500:
    lats.append(elem[1].latitude)
    longs.append(elem[1].longitude)

In [221]:
lats = pd.Series(lats, name='lats')
longs = pd.Series(longs, name='longs')

In [222]:
len(lats)

500

In [237]:
df_first500 = df.loc[0:499,:].copy()

In [238]:
df_first500['latitude'] = lats
df_first500['longitude'] = longs

In [239]:
df_first500.head()

OBJECTID  WOKEY  WO_STATUS  LAST90DAYRPRD  CURRENT_QTR  PREVIOUS_1QTR  \
0         1   2456  COMPLETED              0            0              0   
1         2   2514  COMPLETED              0            0              0   
2         3   2649  COMPLETED              0            0              0   
3         4   1591  COMPLETED              0            0              0   
4         5   1680  COMPLETED              0            0              0   

   PREVIOUS_2QTR  PREVIOUS_3QTR  PREVIOUS_4QTR                        INITDT  \
0              0              0              0  03/29/2010 07:00:00 AM +0000   
1              0              0              0  03/30/2010 07:00:00 AM +0000   
2              0              0              0  03/31/2010 07:00:00 AM +0000   
3              0              0              0  03/15/2010 07:00:00 AM +0000   
4              0              0              0  03/16/2010 07:00:00 AM +0000   

                     FLDSTARTDT                      FLDENDDT  \
0  04/06/2010 07:00:00 AM +0000  04/06/2010 07:00:00 AM +0000   
1  04/01/2010 07:00:00 AM +0000  04/01/2010 07:00:00 AM +0000   
2  04/01/2010 07:00:00 AM +0000  04/01/2010 07:00:00 AM +0000   
3  03/16/2010 07:00:00 AM +0000  03/16/2010 07:00:00 AM +0000   
4  03/17/2010 07:00:00 AM +0000  03/17/2010 07:00:00 AM +0000   

                                            LOCATION  \
0  Holden pl SW bt  SW Kenyon Holden             ...   
1  NE Pacific Street, one block west of Montlake ...   
2  32nd Ave NE and  NE 135th                     ...   
3  Charles st & 6 Ave S pothole next to traffic l...   
4  N 80th St between Corliss Ave N and 2nd Ave NE...   

                                            ADDRDESC  SHAPE   latitude  \
0  HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYO...    NaN  47.531858   
1  MONTLAKE BLVD NE AND NE PACIFIC ST            ...    NaN  47.649014   
2  NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY ...    NaN  47.707446   
3  S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE...    NaN  47.597085   
4  N 80TH ST BETWEEN CORLISS (W LEG) AVE N AND CO...    NaN  47.606209   

    longitude  
0 -122.363186  
1 -122.304655  
2 -122.301566  
3 -122.325065  
4 -122.332071

In [243]:
import cPickle as pickle

In [245]:
df_first500.to_pickle('df_first500.pkl')

In [246]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(500,1000):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc, timeout=10)))
    return new_locs

In [247]:
geocodes_501to1000 = get_geocodes(df)

In [248]:
geocodes_501to1000

[(500, Location((47.5579035, -122.3134718, 0.0))),
 (501, Location((47.6208963, -122.3430313, 0.0))),
 (502, Location((47.5228588, -122.3688324, 0.0))),
 (503, Location((47.5366644, -122.3765918, 0.0))),
 (504, Location((47.6555592, -122.3706272, 0.0))),
 (505, Location((47.5355392, -122.3704759, 0.0))),
 (506, Location((47.6062095, -122.3320708, 0.0))),
 (507, Location((47.6176056, -122.3074067, 0.0))),
 (508, Location((47.6208963, -122.3430313, 0.0))),
 (509, Location((47.6062095, -122.3320708, 0.0))),
 (510, Location((47.6062095, -122.3320708, 0.0))),
 (511, Location((47.5392151, -122.2995882, 0.0))),
 (512, Location((47.6103624, -122.3075774, 0.0))),
 (513, Location((47.7158903, -122.3288386, 0.0))),
 (514, Location((47.6062095, -122.3320708, 0.0))),
 (515, Location((47.6920814, -122.3683148, 0.0))),
 (516, Location((47.6062095, -122.3320708, 0.0))),
 (517, Location((47.6176056, -122.3074067, 0.0))),
 (518, Location((47.5837022, -122.3141117, 0.0))),
 (519, Location((47.6062095, -1

In [285]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(1000,1500):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc, timeout=10)))
    return new_locs

In [286]:
geocodes_1001to1500 = get_geocodes(df)

In [287]:
def get_geocodes(df):
    new_locs = []
    for row in xrange(1500,2000):
        loc = df.ix[row,'ADDRDESC'] + ' Seattle'
        new_locs.append((row, geolocator.geocode(loc, timeout=10)))
    return new_locs

In [288]:
geocodes_1501to2000 = get_geocodes(df)

In [289]:
geocodes_1to2000 = geocodes_1to500 + geocodes_501to1000 + geocodes_1001to1500 + geocodes_1501to2000

####These are geocodes for the first 2000 records.  Create df_first2000.  Then get reverse geocodes for records 501 to 2000, except that will exceed my quota for today.

In [290]:
#Create Series of lats, longs
lats, longs = [], []
for elem in geocodes_1to2000:
    lats.append(elem[1].latitude)
    longs.append(elem[1].longitude)
    
lats = pd.Series(lats, name='lats')
longs = pd.Series(longs, name='longs')

#Copy df
df_first2000 = df.loc[0:1999,:].copy()

df_first2000['latitude'] = lats
df_first2000['longitude'] = longs

#Extract non-specific addresses that are part of Location object and add to dataframe
addrs = []
for elem in geocodes_1to2000:
    addrs.append(elem[1].address)
addrs = pd.Series(addrs, name='addrs')
df_first2000['address'] = addrs


In [291]:
#Reverse geocode lat-longs. I only need to do so for the last 1500.  But I don't think I have room in my quota for today.
#Try 500.
def reverse_geocodes(df):
    rev_locs = []
    for row in xrange(500,2000,1):
        rev_loc = df.ix[row,'latitude'], df.ix[row,'longitude']
        addr = geolocator.reverse(rev_loc, exactly_one=True, timeout=10).address
        rev_locs.append((row, addr))
    return rev_locs

In [292]:
rev_geocodes_500to2000_addrs = reverse_geocodes(df_first2000) #Didn't work.  Quota exceeded.  Tomorrow, do all 1500

GeocoderQuotaExceeded: The given key has gone over the requests limit in the 24 hour period or has submitted too many requests in too short a period of time.

In [ ]:
#Create Series of specific addresses from reverse geocoding and add to df
addrs_dets = []
for elem in rev_geocodes_1to500_addrs:
    addrs_dets.append(elem[1])
addrs_dets = pd.Series(addrs_dets, name='addrs_det')

df_first2000['address_detail'] = addrs_dets

####Do I have the address?

In [251]:
geocodes_1to500[0]

(0, Location((47.5318578, -122.3631857, 0.0)))

In [252]:
geocodes_1to500[0][0]

0

In [253]:
geocodes_1to500[0][1]

Location((47.5318578, -122.3631857, 0.0))

In [254]:
geocodes_1to500[0][1].latitude

47.5318578

In [255]:
geocodes_1to500[0][1].longitude

-122.3631857

In [256]:
geocodes_1to500[0][1].address

u'SW Kenyon St, Seattle, WA, USA'

In [257]:
geocodes_1to500[1][1].address

u'Montlake Blvd NE & NE Pacific St, Seattle, WA 98105, USA'

In [258]:
geocodes_1to500[2][1].address

u'Lake City Way NE, Seattle, WA, USA'

In [259]:
geocodes_1to500[3][1].address

u'Maynard Ave S, Seattle, WA, USA'

####Add first 500 addresses to df

In [260]:
addrs = []
for elem in geocodes_1to500:
    addrs.append(elem[1].address)
addrs = pd.Series(addrs, name='addrs')
df_first500['address'] = addrs

In [261]:
df_first500.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 18 columns):
OBJECTID         500 non-null int64
WOKEY            500 non-null int64
WO_STATUS        500 non-null object
LAST90DAYRPRD    500 non-null int64
CURRENT_QTR      500 non-null int64
PREVIOUS_1QTR    500 non-null int64
PREVIOUS_2QTR    500 non-null int64
PREVIOUS_3QTR    500 non-null int64
PREVIOUS_4QTR    500 non-null int64
INITDT           500 non-null object
FLDSTARTDT       494 non-null object
FLDENDDT         494 non-null object
LOCATION         500 non-null object
ADDRDESC         500 non-null object
SHAPE            0 non-null float64
latitude         500 non-null float64
longitude        500 non-null float64
address          500 non-null object
dtypes: float64(3), int64(8), object(7)
memory usage: 74.2+ KB


In [266]:
df_first500[['ADDRDESC','latitude','longitude','address']].head(50)

ADDRDESC   latitude   longitude  \
0   HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYO...  47.531858 -122.363186   
1   MONTLAKE BLVD NE AND NE PACIFIC ST            ...  47.649014 -122.304655   
2   NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY ...  47.707446 -122.301566   
3   S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE...  47.597085 -122.325065   
4   N 80TH ST BETWEEN CORLISS (W LEG) AVE N AND CO...  47.606209 -122.332071   
5   18TH AVE NE BETWEEN NE 52ND ST AND NE 55TH ST ...  47.668522 -122.295630   
6   3RD AVE NW BETWEEN NW 125TH ST AND NW 127TH ST...  47.721476 -122.364984   
7   SW KENYON ST BETWEEN 15TH AVE SW AND 16TH AVE ...  47.515868 -122.355144   
8   35TH AVE SW AND SW JUNEAU ST                  ...  47.550130 -122.376353   
9   MINOR AVE N AND REPUBLICAN ST                 ...  47.623131 -122.333003   
10  17TH AVE W BETWEEN W DRAVUS ST AND W BERTONA S...  47.650311 -122.388604   
11  DAYTON AVE N AND N 104TH ST                   ...  47.704361 -122.352763   
12  PIKE ST BETWEEN 2ND AVE AND 3RD AVE           ...  47.610741 -122.338703   
13  NE 45TH ST BETWEEN 15TH AVE NE AND 16TH AVE NE...  47.675800 -122.310849   
14  HILLSIDE DR E AND LAKE WASHINGTON BLVD E      ...  47.622480 -122.284664   
15  2ND AVE S BETWEEN S SPOKANE (S RD) ST AND S AN...  47.568293 -122.301841   
16  NE 92ND ST BETWEEN 1ST AVE NE AND 2ND AVE NE  ...  47.708664 -122.327308   
17  AURORA AVE N AND N 103RD (S LEG) ST           ...  47.606209 -122.332071   
18  N 85TH ST BETWEEN NESBIT AVE N AND MIDVALE AVE...  47.729026 -122.343668   
19  N 105TH ST BETWEEN N NORTHGATE (E LEG) WAY AND...  47.730551 -122.338184   
20  JONES AVE NW BETWEEN NW 87TH ST AND NW 89TH PL...  47.693551 -122.385820   
21  S ELMGROVE ST BETWEEN 5TH AVE S AND 7TH AVE S ...  47.561221 -122.324266   
22  GREENWOOD AVE N BETWEEN N 103RD ST AND N 104TH...  47.704361 -122.352763   
23  14TH AVE NW BETWEEN NW 59TH ST AND NW 60TH ST ...  47.672365 -122.381132   
24  RAINIER AVE S BETWEEN 51ST AVE S AND STURTEVAN...  47.518237 -122.268508   
25  S LANE ST BETWEEN MAYNARD AVE S AND 7TH AVE S ...  47.561221 -122.324266   
26  24TH AVE E BETWEEN E MILLER ST AND EAST MONTLA...  47.643274 -122.303345   
27  28TH AVE S BETWEEN S MASSACHUSETTS ST AND S GR...  47.586939 -122.305226   
28  5TH AVE S BETWEEN S JACKSON ST AND S KING ST  ...  47.598329 -122.311574   
29  3RD AVE NW BETWEEN NW 86TH ST AND NW 87TH (S L...  47.606209 -122.332071   
30  S DIRECTOR ST BETWEEN 7TH (E LEG) AVE S AND 8T...  47.573286 -122.322591   
31  SW CAMBRIDGE ST BETWEEN 21ST AVE SW AND KELSEY...  47.519192 -122.361104   
32  NE PACIFIC ST BETWEEN UNIVERSITY WAY NE AND 15...  47.714049 -122.312501   
33  32ND AVE NW AND NW 68TH (S LEG) ST            ...  47.606209 -122.332071   
34  SYLVAN WAY SW BETWEEN SW SYLVAN HEIGHTS DR AND...  47.540225 -122.368392   
35  LAKE CITY WAY NE BETWEEN NE 140TH ST AND NE 14...  47.732017 -122.301944   
36  E MERCER ST BETWEEN 15TH (N LEG) AVE E AND 16T...  47.630397 -122.311242   
37  22ND AVE S BETWEEN S HILL ST AND S WALKER ST  ...  47.583702 -122.314112   
38  NW 140TH ST BETWEEN 2ND AVE NW AND 3RD AVE NW ...  47.716002 -122.360934   
39  E HAMLIN ST BETWEEN MONTLAKE BLVD E AND 24TH A...  47.632286 -122.301604   
40  37TH AVE S BETWEEN S ANDOVER ST AND S DAKOTA S...  47.565993 -122.297900   
41  6TH AVE BETWEEN OLIVE WAY AND STEWART ST      ...  47.614951 -122.335297   
42  NE 133RD ST BETWEEN 20TH AVE NE AND 22ND AVE N...  47.719357 -122.305905   
43  W CREMONA ST BETWEEN QUEEN ANNE AVE N AND 3RD ...  47.635962 -122.360952   
44  BEACON (E RD) AVE S BETWEEN S MYRTLE ST AND S ...  47.537023 -122.293035   
45  NW MARKET ST BETWEEN 24TH AVE NW AND 26TH AVE ...  47.683119 -122.389789   
46  ALASKAN WAY BETWEEN LENORA ST AND BLANCHARD ST...  47.614923 -122.342216   
47  NE 140TH ST BETWEEN 1ST AVE NE AND 2ND AVE NE ...  47.708664 -122.327308   
48  BATTERY ST BETWEEN WESTERN AVE AND 1ST AVE    ...  47.611501 -122.343350   
49  HARVARD AVE E BETWEEN E REPUBLICAN ST AND

####This also provides a diagnostic on quality of geocoding.  Later, clean ADDRDESC to get better geocodes.
####Need to associate socio-economics with address 

In [268]:
df_first500.ix[0,'latitude'], df_first500.ix[0,'longitude']

(47.531857799999997, -122.3631857)

In [271]:
rev_loc = df_first500.ix[0,'latitude'], df_first500.ix[0,'longitude']
geolocator.reverse(rev_loc, exactly_one=True, timeout=10).address

u'2438 SW Kenyon St, Seattle, WA 98106, USA'

####This seems to work.  Now try for the first 500.  Actual API calls.

In [272]:
def reverse_geocodes(df):
    rev_locs = []
    for row in xrange(500):
        rev_loc = df.ix[row,'latitude'], df.ix[row,'longitude']
        addr = geolocator.reverse(rev_loc, exactly_one=True, timeout=10).address
        rev_locs.append((row, addr))
    return rev_locs

In [273]:
rev_geocodes_1to500_addrs = reverse_geocodes(df_first500)

In [274]:
rev_geocodes_1to500_addrs[:10]

[(0, u'2438 SW Kenyon St, Seattle, WA 98106, USA'),
 (1, u'3705-3707 Montlake Blvd NE, Seattle, WA 98105, USA'),
 (2, u'10754-10850 Lake City Way NE, Seattle, WA 98125, USA'),
 (3, u'608 Maynard Ave S, Seattle, WA 98104, USA'),
 (4, u'909 5th Ave, Seattle, WA 98164, USA'),
 (5, u'2914 NE 55th St, Seattle, WA 98105, USA'),
 (6, u'12701 7th Ave NW, Seattle, WA 98177, USA'),
 (7, u'9767 16th Ave SW, Seattle, WA 98106, USA'),
 (8, u'3503 SW Juneau St, Seattle, WA 98126, USA'),
 (9, u'500 Minor Ave N, Seattle, WA 98109, USA')]

####This seemed to work.  Add 'address_detailed' to df

In [277]:
addrs_dets = []
for elem in rev_geocodes_1to500_addrs:
    addrs_dets.append(elem[1])
addrs_dets = pd.Series(addrs_dets, name='addrs_det')


In [280]:
df_first500['address_detail'] = addrs_dets

In [281]:
df_first500[['ADDRDESC','latitude','longitude','address','address_detail']].head(50)

ADDRDESC   latitude   longitude  \
0   HOLDEN PL SW BETWEEN SW HOLDEN ST AND SW KENYO...  47.531858 -122.363186   
1   MONTLAKE BLVD NE AND NE PACIFIC ST            ...  47.649014 -122.304655   
2   NE 135TH ST BETWEEN 32ND AVE NE AND LAKE CITY ...  47.707446 -122.301566   
3   S CHARLES ST BETWEEN 6TH AVE S AND MAYNARD AVE...  47.597085 -122.325065   
4   N 80TH ST BETWEEN CORLISS (W LEG) AVE N AND CO...  47.606209 -122.332071   
5   18TH AVE NE BETWEEN NE 52ND ST AND NE 55TH ST ...  47.668522 -122.295630   
6   3RD AVE NW BETWEEN NW 125TH ST AND NW 127TH ST...  47.721476 -122.364984   
7   SW KENYON ST BETWEEN 15TH AVE SW AND 16TH AVE ...  47.515868 -122.355144   
8   35TH AVE SW AND SW JUNEAU ST                  ...  47.550130 -122.376353   
9   MINOR AVE N AND REPUBLICAN ST                 ...  47.623131 -122.333003   
10  17TH AVE W BETWEEN W DRAVUS ST AND W BERTONA S...  47.650311 -122.388604   
11  DAYTON AVE N AND N 104TH ST                   ...  47.704361 -122.352763   
12  PIKE ST BETWEEN 2ND AVE AND 3RD AVE           ...  47.610741 -122.338703   
13  NE 45TH ST BETWEEN 15TH AVE NE AND 16TH AVE NE...  47.675800 -122.310849   
14  HILLSIDE DR E AND LAKE WASHINGTON BLVD E      ...  47.622480 -122.284664   
15  2ND AVE S BETWEEN S SPOKANE (S RD) ST AND S AN...  47.568293 -122.301841   
16  NE 92ND ST BETWEEN 1ST AVE NE AND 2ND AVE NE  ...  47.708664 -122.327308   
17  AURORA AVE N AND N 103RD (S LEG) ST           ...  47.606209 -122.332071   
18  N 85TH ST BETWEEN NESBIT AVE N AND MIDVALE AVE...  47.729026 -122.343668   
19  N 105TH ST BETWEEN N NORTHGATE (E LEG) WAY AND...  47.730551 -122.338184   
20  JONES AVE NW BETWEEN NW 87TH ST AND NW 89TH PL...  47.693551 -122.385820   
21  S ELMGROVE ST BETWEEN 5TH AVE S AND 7TH AVE S ...  47.561221 -122.324266   
22  GREENWOOD AVE N BETWEEN N 103RD ST AND N 104TH...  47.704361 -122.352763   
23  14TH AVE NW BETWEEN NW 59TH ST AND NW 60TH ST ...  47.672365 -122.381132   
24  RAINIER AVE S BETWEEN 51ST AVE S AND STURTEVAN...  47.518237 -122.268508   
25  S LANE ST BETWEEN MAYNARD AVE S AND 7TH AVE S ...  47.561221 -122.324266   
26  24TH AVE E BETWEEN E MILLER ST AND EAST MONTLA...  47.643274 -122.303345   
27  28TH AVE S BETWEEN S MASSACHUSETTS ST AND S GR...  47.586939 -122.305226   
28  5TH AVE S BETWEEN S JACKSON ST AND S KING ST  ...  47.598329 -122.311574   
29  3RD AVE NW BETWEEN NW 86TH ST AND NW 87TH (S L...  47.606209 -122.332071   
30  S DIRECTOR ST BETWEEN 7TH (E LEG) AVE S AND 8T...  47.573286 -122.322591   
31  SW CAMBRIDGE ST BETWEEN 21ST AVE SW AND KELSEY...  47.519192 -122.361104   
32  NE PACIFIC ST BETWEEN UNIVERSITY WAY NE AND 15...  47.714049 -122.312501   
33  32ND AVE NW AND NW 68TH (S LEG) ST            ...  47.606209 -122.332071   
34  SYLVAN WAY SW BETWEEN SW SYLVAN HEIGHTS DR AND...  47.540225 -122.368392   
35  LAKE CITY WAY NE BETWEEN NE 140TH ST AND NE 14...  47.732017 -122.301944   
36  E MERCER ST BETWEEN 15TH (N LEG) AVE E AND 16T...  47.630397 -122.311242   
37  22ND AVE S BETWEEN S HILL ST AND S WALKER ST  ...  47.583702 -122.314112   
38  NW 140TH ST BETWEEN 2ND AVE NW AND 3RD AVE NW ...  47.716002 -122.360934   
39  E HAMLIN ST BETWEEN MONTLAKE BLVD E AND 24TH A...  47.632286 -122.301604   
40  37TH AVE S BETWEEN S ANDOVER ST AND S DAKOTA S...  47.565993 -122.297900   
41  6TH AVE BETWEEN OLIVE WAY AND STEWART ST      ...  47.614951 -122.335297   
42  NE 133RD ST BETWEEN 20TH AVE NE AND 22ND AVE N...  47.719357 -122.305905   
43  W CREMONA ST BETWEEN QUEEN ANNE AVE N AND 3RD ...  47.635962 -122.360952   
44  BEACON (E RD) AVE S BETWEEN S MYRTLE ST AND S ...  47.537023 -122.293035   
45  NW MARKET ST BETWEEN 24TH AVE NW AND 26TH AVE ...  47.683119 -122.389789   
46  ALASKAN WAY BETWEEN LENORA ST AND BLANCHARD ST...  47.614923 -122.342216   
47  NE 140TH ST BETWEEN 1ST AVE NE AND 2ND AVE NE ...  47.708664 -122.327308   
48  BATTERY ST BETWEEN WESTERN AVE AND 1ST AVE    ...  47.611501 -122.343350   
49  HARVARD AVE E BETWEEN E REPUBLICAN ST AND

In [282]:
df_first500[['ADDRDESC','latitude','longitude','address','address_detail']].tail(50)

ADDRDESC   latitude   longitude  \
450  11TH AVE BETWEEN E OLIVE ST AND E HOWELL ST   ...  47.617606 -122.307407   
451  ALOHA ST BETWEEN 2ND AVE N AND 3RD AVE N      ...  47.636026 -122.351942   
452  30TH AVE S BETWEEN S MCCLELLAN ST AND S MOUNT ...  47.606209 -122.332071   
453  NW 96TH ST BETWEEN 23RD AVE NW AND 24TH AVE NW...  47.684657 -122.387589   
454  3RD AVE W BETWEEN W BERTONA ST AND W NICKERSON...  47.653279 -122.369132   
455  8TH AVE N BETWEEN THOMAS ST AND HARRISON ST   ...  47.622089 -122.343036   
456  N 88TH ST BETWEEN LINDEN AVE N AND AURORA (N L...  47.606209 -122.332071   
457  HARVARD AVE BETWEEN E PIKE ST AND E PINE ST   ...  47.615290 -122.304892   
458  W EWING ST BETWEEN 6TH AVE W AND W EWING PL   ...  47.655559 -122.370627   
459  1ST AVE NE BETWEEN CORLISS AVE N AND N 128TH S...  47.722057 -122.339568   
460  JAMES ST BETWEEN 6TH AVE AND 7TH AVE          ...  47.607480 -122.329575   
461  33RD AVE S BETWEEN S OREGON (E LEG) ST AND REN...  47.520259 -122.272693   
462  PERKINS LN W BETWEEN DEAD END AND W RAYE ST   ...  47.643468 -122.387632   
463  22ND AVE E BETWEEN E HIGHLAND DR AND E GALER S...  47.632287 -122.302300   
464  RAINIER AVE S BETWEEN S CLOVERDALE ST AND S HE...  47.523086 -122.311779   
465  COLORADO AVE S BETWEEN S HINDS ST AND S SPOKAN...  47.606209 -122.332071   
466  BOREN AVE S BETWEEN 12TH AVE S AND S MAIN ST  ...  47.600008 -122.311577   
467  4TH AVE S BETWEEN S HENDERSON (S LEG) ST AND S...  47.521386 -122.287151   
468  2ND AVE S BETWEEN DIAGONAL AVE S AND DEAD END ...  47.606209 -122.332071   
469  E OLIVE ST BETWEEN 13TH AVE AND 14TH AVE      ...  47.610383 -122.314140   
470  NW MARKET ST BETWEEN NW 54TH ST AND 30TH AVE N...  47.680941 -122.395704   
471  MERIDIAN AVE N BETWEEN N 105TH ST AND N 106TH ...  47.705969 -122.332569   
472  EVANSTON AVE N BETWEEN N 117TH ST AND N 120TH ...  47.715943 -122.339496   
473  ALOHA ST BETWEEN WARREN AVE N AND 2ND AVE N   ...  47.632701 -122.353203   
474  SW ROXBURY ST BETWEEN 27TH AVE SW AND 28TH AVE...  47.531882 -122.368532   
475  23RD AVE BETWEEN E CHERRY ST AND E COLUMBIA ST...  47.609018 -122.300828   
476  E JEFFERSON ST BETWEEN BROADWAY AND 10TH AVE  ...  47.606207 -122.319427   
477  NE 65TH ST BETWEEN 58TH AVE NE AND SAND POINT ...  47.693053 -122.271752   
478  1ST AVE NW BETWEEN NW 125TH ST AND NW 127TH ST...  47.721476 -122.364984   
479  15TH AVE W BETWEEN 15TH (NB OFF RP) AVE W AND ...  47.650311 -122.388604   
480  N 127TH ST BETWEEN PHINNEY AVE N AND DAYTON AV...  47.714813 -122.352861   
481  JAMES ST BETWEEN 3RD AVE AND 4TH AVE          ...  47.610800 -122.337217   
482  23RD AVE BETWEEN E FIR ST AND E SPRUCE ST     ...  47.603510 -122.303735   
483  E BOSTON ST BETWEEN 11TH AVE E AND 12TH AVE E ...  47.628440 -122.316775   
484  EARL AVE NW BETWEEN NW 88TH ST AND NW 90TH ST ...  47.694685 -122.378428   
485  43RD AVE NE BETWEEN NE 38TH ST AND NE 41ST ST ...  47.658628 -122.290708   
486  23RD AVE BETWEEN E TERRACE ST AND E JEFFERSON ...  47.606171 -122.303803   
487  E VALLEY ST BETWEEN 26TH AVE E AND 28TH AVE E ...  47.626641 -122.296394   
488  S HOLLY ST BETWEEN 46TH AVE S AND RAINIER AVE ...  47.544847 -122.274497   
489  INTERLAKE AVE N BETWEEN N 137TH ST AND N 140TH...  47.606209 -122.332071   
490  LAKE PARK DR S BETWEEN SHORELAND DR S AND S MC...  47.578355 -122.301287   
491  28TH AVE S BETWEEN S WALKER ST AND S COLLEGE S...  47.582954 -122.305583   
492  27TH AVE NE BETWEEN NE 97TH ST AND NE 98TH ST ...  47.700282 -122.303333   
493  S DAWSON ST BETWEEN 37TH AVE S AND 39TH AVE S ...  47.534773 -122.283611   
494  S JACKSON ST BETWEEN 1ST AVE S AND OCCIDENTAL ...  47.587816 -122.333248   
495  SW BARTON ST BETWEEN 35TH AVE SW AND 36TH AVE ...  47.542885 -122.377814   
496  2ND AVE S BETWEEN S SPOKANE (S RD) ST AND S AN...  47.568293 -122.301841   
497  E SHELBY ST BETWEEN BOYER AVE E AND DEAD END 3...  47.606209 -122.332071   
498  NE 63RD ST BETWEEN 16TH AVE NE AND 17TH AVE NE...  47.699426 -122.309494

####Overwrite the pickle

In [283]:
df_first500_addr = df_first500.copy()

In [284]:
df_first500_addr.to_pickle('df_first500_addr.pkl')

This updated pickled df has geocodes, address, and detailed address for the first 500 rows.  In some cases, geocoding fails and only returns Seattle, WA.

Ran into quota limits again.  Need to solve this problem tomorrow, either: (1) Work with multiple keys and/or (2) go to AWS

####Let's try Nominatim again.  It has no query limits.  If it's quality is ok, then go with it.

In [293]:
geolocator2 = Nominatim()

In [294]:
location = geolocator2.geocode("market and 4th san francisco")

In [295]:
print(location.address)

San Francisco 4th & King Street Station, 4th Street, South Beach, SF, California, 94107, United States of America


In [296]:
print((location.latitude, location.longitude))

(37.7765767, -122.394827912059)


In [297]:
print(location.raw)

{u'display_name': u'San Francisco 4th & King Street Station, 4th Street, South Beach, SF, California, 94107, United States of America', u'importance': 0.61490776660872, u'place_id': u'61306801', u'lon': u'-122.394827912059', u'lat': u'37.7765767', u'osm_type': u'way', u'licence': u'Data \xa9 OpenStreetMap contributors, ODbL 1.0. http://www.openstreetmap.org/copyright', u'osm_id': u'28295229', u'boundingbox': [u'37.7762267', u'37.776876', u'-122.3951705', u'-122.3945376'], u'type': u'station', u'class': u'railway', u'icon': u'https://nominatim.openstreetmap.org/images/mapicons/transport_train_station2.p.20.png'}


In [301]:
location = geolocator2.geocode('SW HOLDEN ST AND SW KENYON ST seattle WA')

In [302]:
print location.latitude, location.longitude

AttributeError: 'NoneType' object has no attribute 'latitude'

In [303]:
location = geolocator2.geocode('montlake blvd NE and NE pacific st seattle')

In [304]:
print location.latitude, location.longitude

AttributeError: 'NoneType' object has no attribute 'latitude'

It doesn't work on examples that GoogleV3 works on.

####Need to create most basic of features, a categorical variable on neighborhood that the pothole falls in.  I might need to do that in a separate notebook.  Yes.  I can use the same approach to get the socio-economics, hopefully.

####Also, Ben suggested doing NLP on the "location" column.